fn/async.go

```diff
@@ -0,0 +1,110 @@
+package fn
+
+import (
+	"context"
+	"errors"
+)
+
+// Async represents an action whose results are being computed outside this
+// control thread.
+type Async[A any] interface {
+	// Await may be called to block on the completion of the remote
+	// computation.
+	Await() (*A, error)
+
+	// Cancel may be called to abandon the remote computation.
+	Cancel()
+}
+
+var (
+	// ContextCancelled indicates that this Async was cancelled because the
+	// surrounding context was being torn down.
+	ContextCancelled = errors.New("context cancelled")
+)
+
+// RunAsyncWithCancel is a function that will transform a synchronous function
+// and make it a cancellable asynchronous call. This can be useful if the
+// function in the argument can block indefinitely and the calling context needs
+// a way to abandon waiting for it to complete.
+func RunAsyncWithCancel[A any](ctx context.Context,
+	f func(ctx context.Context) (*A, error)) Async[A] {
+
+	localCtx, cancel := context.WithCancel(ctx)
+
+	a := async[A]{
+		ctx:    localCtx,
+		cancel: cancel,
+		res:    make(chan *A, 1),
+		err:    make(chan error, 1),
+	}
+
+	go func() {
+		res, err := f(localCtx)
+
+		// If the error we get back from f is non-nil we try to send an
+		// err back on the err channel, bailing if we're being
+		// cancelled.
+		if err != nil {
+			select {
+			case a.err <- err:
+			case <-a.ctx.Done():
```

Comment:
Not an expert but feels designwise a bit odd to always check for the cancel channel and not summarize the cases.

Reply:
I'm not sure what you mean here. I made this choice since we don't want to spend effort on sending back over the error channel if we've already been cancelled. I suppose that since the channel is buffered and we only ever send on it once, that this shouldn't ever happen though.

Reply:
I meant the two select statement could be maybe summarized right. Or even go with the proposal of elle which did not use the Done() channel at all because we know they are buffered and will not hold up the process.

```
a.wg.Add(1)
	go func() {
		defer a.wg.Done()

		res, err := f(ctxc)
		if err != nil {
			a.err <- err

			return
		}

		a.res <- res
	}()
```